Test buildpack on Travis

edmorley opened the pull request:
Since previously the Travis script step was a no-op. This will give us confidence that the upcoming Redis version support PR has not broken anything.

Fixes [W-6165951](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000006tn5eIAA/view).

edmorley committed:
Test buildpack on Travis

Since previously the Travis script step was a no-op. This will give us
confidence that the upcoming Redis version support PR has not broken
anything.

Fixes W-6165951.

bensymonds commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+FROM $BUILD_IMAGE as builder
+
+ARG REDIS_VERSION
```
I can't figure out what this does.

edmorley replied:
Oh sorry that was one line left from testing the approach would work ok for Redis version support (the next PR)

edmorley committed:
Remove stray redis version reference

bensymonds commented on bin/test.sh:
```diff
@@ -0,0 +1,36 @@
+# https://stackoverflow.com/questions/38271281/can-i-use-redis-cli-with-a-connection-url
+REDIS_CONNECTION_ARGS="echo \${REDIS_URL} | sed 's_redis://\(.*\):\(.*\)@\(.*\):\(.*\)_-h \3 -p \4 -a \2_'"
+TEST_COMMAND="source .profile.d/redis.sh && redis-cli \$(${REDIS_CONNECTION_ARGS}) ping"
+docker run --rm -it "${OUTPUT_IMAGE}" bash -c "${TEST_COMMAND}"
```
Just to double-check, are you confident the correct exit code ends up coming out here? Just thinking:

* the `docker run` needs to return the exit code from the `bash -c ...` bit.
* the `bash -c` bit needs to return the exit code of `TEST_COMMAND`
* `redis-cli` needs to return a non-zero exit code if the connection fails.
* (possibly missed some steps, you see what I mean though)

I have to ask just because obviously it's quite important for this script to work as a CI thing!

edmorley replied:
Testing this locally, the exit code works for everything but redis-cli, which led me to:
https://github.com/antirez/redis/issues/2270#issuecomment-76777000

I'm hopeful Redis 4/5 has fixed this, in which case we'll have coverage from those parts of the test matrix. If they haven't, I'll update this to do a grep for PONG instead.

edmorley replied:
The exit code works fine in Redis 4 + 5 so going to leave this as-is for now (with the next PR we'll be testing against multiple versions, and REDIS_URL is set the same for all of the, so some coverage is fine). Hopefully we'll drop support for Redis 3 at some point in the future too.

bensymonds replied:
I don't quite follow. So if the Redis version is 3 then the script may fail silently? That doesn't sound very good.

bensymonds commented on Dockerfile:
```diff
@@ -0,0 +1,17 @@
+RUN /buildpack/bin/detect /build
+RUN /buildpack/bin/compile /build /cache /env
+
+
```
Might want to clean up the extra blank line. (Don't hit me.)

bensymonds replied:
Actually I just realised it might be intentional, to separate the two `FROM` sections. Maybe we should add another one above the first `FROM` then. So the sections are:

1) General ARGs (shared throughout following sections)
2) `FROM $BUILD_IMAGE as builder`
3) `FROM $RUNTIME_IMAGE`

edmorley replied:
Yeah. Agreed, I can do that or else code comments before both, though the titles are pretty much going to repeat the obvious :-)

bensymonds replied:
Cool yeah I agree comments that repeat the obvious are not good. I guess a third option would just be `#############` as a separator! At least then the comment can't go out of date :smile: But meh, not a big deal.

bensymonds reviewed (approved):
Made a couple of comments but apart from those it looks good :+1:

edmorley committed:
Clean up REDIS_URL parsing, add .dockerignore, remove bashrc addition

The REDIS_URL parsing change is since Redis 4 is because it previously
included a trailing slash, which Redis 3 is fine with, but Redis 4 is not.

The bashrc removal is since it's not required now that the file is
sourced in test.sh.

edmorley committed:
Add comments to explain the multi-stage build

Also resolves the whitespace inconsistency.
